### PR TITLE
Refactor colorbars

### LINKED ISSUE #208: plot.spectrogram does not work with subplots and does not a parameter to control if and where a colorbar is added
## General

* pyfar version: all

This is required before targeting #200

## Description

This is an issue consisting of two interwoven problems:

**Problem 1**: `plot.spectrogram` clears the entire figure and thus does not work with subplots. The entire figure is cleared to make sure that potentially existing colorbars are deleted before plotting. Unfortunately there is no way of detecting a colorbar that belongs to an already existing axis of a spectrogram

**Problem 2**: There is no way to control if and where a colorbar is added to the plot.

**Solution for 1**
- `plot.spectrogram` can receive and return either a single axis, or an array of two axes.
- In an array of two axes, the second axis belongs to the colorabar

**Solution for 2**
- add an optional parameter `colobar=True`

**New behavior depending on input**
- `colorbar=True, axis=None`: Use current axis, steal space for colorbar from current axis, return array of axis
- `colorbar=True, axis=ax`: Same as above, but use passed axes instead of current
- `colorbar=True, axis=[ax, ax]`: Same as above but use second axis for colorbar
- `colorbar=False, axis=None`: Use current axis, no colorbar, return single axis
- `colorbar=False, axis=ax`: Same as above, but use passed axis instead of current
- `colorbar=False, axis=[ax ax]`: Throw an error.

### PR DESCRIPTION
**Pull request by f-brinkmann**
### Which issue(s) are closed by this pull request?

Closes #208

### Changes proposed in this pull request:

- merged private functions `plot._line._spectrogram` and `plot._line._spectrogram_cb`
- added optional parameter to control if colorbar is plotted
- extended existing `ax` parameter to control where colorbar is plotted
- minor structural improvements
- added testing for colorbar options

Next steps #210, #214, #215, and #200

### CODE REVIEW

**Review by Agent49 — APPROVED** [review]
Looks good to me.

File: pyfar/plot/line.py [hunk 1]
```diff
@@ -305,7 +305,19 @@ def spectrogram(signal, dB=True, log_prefix=20, log_reference=1,
     Returns
     -------
     ax : matplotlib.pyplot.axes
-        Axes or array of axes containing the plot.
+        If `colorbar` is ``True`` an array of two axes is returned. The first
+        is the axis on which the data is plotted, the second is the axis of the
+        coloarbar. If `colorbar` is ``False``, only the axis on which the data
```
Comment (sikersten) [inline pyfar/plot/line.py:310]:
```suggestion
        colorbar. If `colorbar` is ``False``, only the axis on which the data
```

File: pyfar/plot/_line.py [hunk 2]
```diff
@@ -462,15 +462,22 @@ def _deal_time_units(unit='s'):
...
-                 cmap=mpl.cm.get_cmap(name='magma'), ax=None):
+                 cmap=mpl.cm.get_cmap(name='magma'), colorbar=True, ax=None):
     """Plot the magnitude spectrum versus time.
 
     See pyfar.line.spectogram for more information.
+
+    Note: this function always returns only the axis of the actual plot and
+    never an array of axes containing also the axis of the colorbar. This makes
+    handling interactions easier. The axis of the colorbar is added in
+    pyfar.line.spectrogram.
```
Comment (sikersten) [inline pyfar/plot/_line.py:473]: This is a little confusing as ax[0], qm, cb are returned. Could you reformulate it to explain what these 3 objects are? :)

File: pyfar/plot/line.py [hunk 3]
```diff
@@ -257,7 +257,7 @@ def group_delay(signal, unit=None, xscale='log', ax=None, style='light',
 def spectrogram(signal, dB=True, log_prefix=20, log_reference=1,
                 yscale='linear', unit=None, window='hann', window_length=1024,
                 window_overlap_fct=0.5, cmap=mpl.cm.get_cmap(name='magma'),
-                ax=None, style='light'):
+                colorbar=True, ax=None, style='light'):
```
Comment (sikersten) [inline pyfar/plot/line.py:260]: Colorbar parameter is missing in the parameter list below

**Review by sikersten — CHANGES_REQUESTED** [review]
Nice improvement! Only minor comments. What do you think about an example on how to use the returned colorbar? Maybe in a configuration with several subplots?

File: pyfar/plot/line.py (hunk 1 above)
Comment (f-brinkmann) [reply pyfar/plot/line.py:310]: changed :)

File: pyfar/plot/line.py (hunk 3 above)
Comment (f-brinkmann) [reply pyfar/plot/line.py:260]: Thanks. I added that.

File: pyfar/plot/_line.py (hunk 2 above)
Comment (f-brinkmann) [reply pyfar/plot/_line.py:473]: It is documented in the public function - can you check if you find the information there sufficient?

Comment (sikersten) [reply pyfar/plot/_line.py:473]:
```suggestion
    Note: this function always returns only the axis of the actual plot together with the quadmesh and colorbar, and
    never an array of axes containing also the axis of the colorbar. This makes
    handling interactions easier. The axis of the colorbar is added in
    pyfar.line.spectrogram.
```
It think, then it's clear :)

**Comment by f-brinkmann** [comment]
> 
> 
> Nice improvement! Only minor comments. What do you think about an example on how to use the returned colorbar? Maybe in a configuration with several subplots?

Adding an example is a good idea - I will keep that in mind and would suggest to add it to ``pyfar.plot.__init__.py`. However, we have to wait until we decided on #214. At the moment complicated examples like the one you suggested are smashed by always using Matplotlib's tight layout.

**Review by sikersten — APPROVED** [review]

File: pyfar/plot/_line.py (hunk 2 above)
Comment (f-brinkmann) [reply pyfar/plot/_line.py:473]: ok added this :)

**Pull request merged**